src/psyclone/domain/common/transformations/invoke_trans.py

```diff
@@ -0,0 +1,238 @@
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford STFC Daresbury Lab
+
+'''Specialise generic PSyIR representing an invoke call within the
+algorithm layer to a PSyclone algorithm-layer-specific invoke call
+which uses specialised classes.
+
+'''
+from fparser.two.Fortran2003 import Structure_Constructor
+
+from psyclone.psyir.nodes import Call, ArrayReference, CodeBlock
+from psyclone.psyir.symbols import Symbol, TypeSymbol, StructureType, \
+    RoutineSymbol
+from psyclone.domain.common.algorithm import AlgorithmInvokeCall, \
+    KernelFunctor
+from psyclone.psyGen import Transformation
+from psyclone.psyir.transformations import TransformationError
+from psyclone.psyir.frontend.fparser2 import Fparser2Reader
+
+
+class InvokeTrans(Transformation):
+    '''Transform a generic PSyIR representation of an Algorithm-layer
+    invoke call to a PSyclone version with specialised domain-specific
+    nodes.
+
+    :param str invoke_name: the name used to specify an invoke \
+        call. This is an optional argument that defaults to 'invoke'.
+
+    '''
+    def __init__(self, invoke_name="invoke"):
+        self._invoke_name = invoke_name
+
+    @staticmethod
+    def _parse_args(code_block, fp2_node):
+        '''Return the arguments from a Structure Constructor stored as a
+        CodeBlock containing an fparser2 ast.
+
+        :param code_block: the CodeBlock containing a StructureConstructor.
+        :type code_block: :py:class:`psyclone.psyir.nodes.CodeBlock`
+        :param fp2_node: the fparser2 Structure Constructor node.
+        :type fp2_node: \
+            :py:class:`fparser.two.Fortran2003.Structure_Constructor`
+
+        :returns: a list of PSyIR nodes containing the \
+            StructureConstructor arguments.
+        :rtype: :py:class:`psyclone.psyir.nodes.node.ChildrenList`
+
+        '''
+        dummy_call = Call(RoutineSymbol("dummy"),
+                          parent=code_block.parent)
+        fparser2 = Fparser2Reader()
+        for arg in fp2_node.children[1].children:
+            fparser2.process_nodes(dummy_call, [arg])
+        dummy_call.parent = None
+        args = dummy_call.children
+        return args
```

Comment:
You should remove the 2 previous lines and return dummy_call.pop_all_children()

Reply:
Ooh, very nice.